### ISSUE TITLE
DownloadCache failing with 'Content has invalid size'

### ISSUE THREAD

**Issue by dmikusa-pivotal** [cloudfoundry/java-buildpack#284]
I've been investigating a problem with the Java build pack when hosting my index.yml files on Github.  The DownloadCache is reporting the following.

```
[DownloadCache]                  WARN  Request failure 1, retrying: Content has invalid size.  Was 367, should be 138.
[DownloadCache]                  WARN  Request failure 2, retrying: Content has invalid size.  Was 367, should be 138.
[DownloadCache]                  WARN  Request failure 3, retrying: Content has invalid size.  Was 367, should be 138.
[DownloadCache]                  WARN  Request failure 4, retrying: Content has invalid size.  Was 367, should be 138.
[DownloadCache]                  WARN  Request failure 5, retrying: Content has invalid size.  Was 367, should be 138.
[InternetAvailability]           WARN  Internet availability set to false: Request failed: Content has invalid size.  Was 367, should be 138.
[DownloadCache]                  WARN  Unable to download https://raw.githubusercontent.com/yooture/runtimes/master/index.yml into cache /var/folders/w4/g5kx6n_16330906jj_6nvqcc0000gp/T: Content has invalid size.  Was 367, should be 138.
[Buildpack]                      ERROR Compile failed with exception #<RuntimeError: Unable to find cached file for https://raw.githubusercontent.com/yooture/runtimes/master/index.yml>
```

Upon digging into this, it seems that GH is returning a gzip'd copy of the file and that the content-length head of that is 138 bytes, where as the unzipped file is 367 bytes.  They are the same, except for the encoding.

Here's debug output from the build pack.

```
DownloadCache]                  DEBUG Proxy: , , ,
[DownloadCache]                  DEBUG Adding HTTP options for secure connection
[DownloadCache]                  DEBUG Adding HTTP options for secure connection
[DownloadCache]                  DEBUG HTTP: raw.githubusercontent.com, 443, {:use_ssl=>true}
[DownloadCache]                  DEBUG SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES128-GCM-SHA256
    Session-ID: 00D9AFAC8A1FAD12DA9D9D76340DDE84C286DA6912BDFBF7BE644F8BD27C156B
    Session-ID-ctx:
    Master-Key: 0300CED0F500608DCA7588CE44EFCDE5226A99A9F6FB8BBE949B4786CD75783519044CDAD103CB39A995498540FB0A9D
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 1200 (seconds)
    TLS session ticket:
    0000 - 63 cc 77 4a 00 db 2c 42-2e 8f 76 23 dd a9 ae 53   c.wJ..,B..v#...S
    0010 - 20 bb 2c 81 e6 73 26 1c-e9 fb 11 3e 02 20 cf 95    .,..s&....>. ..
    0020 - a9 35 ec ba f3 09 42 fe-93 2c c4 7a c5 af ca 09   .5....B..,.z....
    0030 - e2 74 b5 7e b8 2b b6 59-98 4d 30 c7 01 1e 62 38   .t.~.+.Y.M0...b8
    0040 - 64 89 46 ae 54 f4 24 9f-22 4a 5c 6f d5 2d 96 07   d.F.T.$."J\o.-..
    0050 - 25 8c f8 2c e9 f3 f4 61-8f 3c 17 63 b5 b9 fc 9d   %..,...a.<.c....
    0060 - b5 fd 8e bf 46 e7 4d 0a-7a ae d6 39 c8 0b 20 5b   ....F.M.z..9.. [
    0070 - f6 d0 b0 f8 d2 4d 0a 61-8a 83 25 9c 76 84 21 96   .....M.a..%.v.!.
    0080 - de aa 62 97 f9 97 07 94-0d 1c 01 f8 ff dc 95 ba   ..b.............
    0090 - 81 c2 ea 5c dc f5 ca 3a-cb b8 4f 45 95 91 58 bc   ...\...:..OE..X.
    00a0 - fe 1d 7a 74 1a ea d3 46-af 5a 57 39 1e 29 a2 0b   ..zt...F.ZW9.)..
    00b0 - 9f ea e7 1b d0 a2 2a 50-78 2e fd 48 06 0e 1e 2f   ......*Px..H.../

    Start Time: 1458578302
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)

[DownloadCache]                  DEBUG Request: /yooture/runtimes/master/index.yml, {"accept-encoding"=>["gzip;q=1.0,deflate;q=0.6,identity;q=0.3"], "accept"=>["*/*"], "user-agent"=>["Ruby"]}
Exception `OpenSSL::SSL::SSLErrorWaitReadable' at /Users/danielmikusa/.rbenv/versions/2.2.3/lib/ruby/2.2.0/openssl/buffering.rb:182 - read would block
[DownloadCache]                  DEBUG Response headers: {"content-security-policy"=>["default-src 'none'; style-src 'unsafe-inline'"], "x-xss-protection"=>["1; mode=block"], "x-frame-options"=>["deny"], "x-content-type-options"=>["nosniff"], "strict-transport-security"=>["max-age=31536000"], "etag"=>["\"e9fa6dccaeda7e47e2c79335fe570f6212dfecbf\""], "content-type"=>["text/plain; charset=utf-8"], "cache-control"=>["max-age=300"], "x-github-request-id"=>["C71B4F21:7CE0:C4D49D4:56F0237E"], "content-encoding"=>["gzip"], "content-length"=>["138"], "accept-ranges"=>["bytes"], "date"=>["Mon, 21 Mar 2016 16:38:22 GMT"], "via"=>["1.1 varnish"], "connection"=>["keep-alive"], "x-served-by"=>["cache-lax1432-LAX"], "x-cache"=>["MISS"], "x-cache-hits"=>["0"], "vary"=>["Authorization,Accept-Encoding"], "access-control-allow-origin"=>["*"], "x-fastly-request-id"=>["2d9facf0aeb9554f6855e7f4914b74da11149d5f"], "expires"=>["Mon, 21 Mar 2016 16:43:22 GMT"], "source-age"=>["0"]}
[DownloadCache]                  DEBUG Response status: 200
[DownloadCache]                  DEBUG Persisting etag: "e9fa6dccaeda7e47e2c79335fe570f6212dfecbf"
[DownloadCache]                  DEBUG Persisting content to /var/folders/w4/g5kx6n_16330906jj_6nvqcc0000gp/T/https%3A%2F%2Fraw.githubusercontent.com%2Fyooture%2Fruntimes%2Fmaster%2Findex.yml.cached
[DownloadCache]                  DEBUG File ---
1.7.0_71: https://github.com/yooture/runtimes/raw/master/server-jre-7u71-linux-x64.tar.gz
1.8.0_45: https://github.com/yooture/runtimes/raw/master/server-jre-8u45-linux-x64.tar.gz
1.8.0_51: https://github.com/yooture/runtimes/raw/master/server-jre-8u51-linux-x64.tar.gz
1.8.0_74: https://github.com/yooture/runtimes/raw/master/server-jre-8u74-linux-x64.tar.gz



[DownloadCache]                  DEBUG Validated content size 367 is 138
Exception `Errno::ENOTDIR' at /Users/danielmikusa/Downloads/java-buildpack/lib/java_buildpack/util/cache/cached_file.rb:65 - Not a directory @ dir_s_rmdir - /var/folders/w4/g5kx6n_16330906jj_6nvqcc0000gp/T/https%3A%2F%2Fraw.githubusercontent.com%2Fyooture%2Fruntimes%2Fmaster%2Findex.yml.cached
Exception `Errno::ENOTDIR' at /Users/danielmikusa/Downloads/java-buildpack/lib/java_buildpack/util/cache/cached_file.rb:65 - Not a directory @ dir_s_rmdir - /var/folders/w4/g5kx6n_16330906jj_6nvqcc0000gp/T/https%3A%2F%2Fraw.githubusercontent.com%2Fyooture%2Fruntimes%2Fmaster%2Findex.yml.etag
Exception `JavaBuildpack::Util::Cache::InferredNetworkFailure' at /Users/danielmikusa/Downloads/java-buildpack/lib/java_buildpack/util/cache/download_cache.rb:309 - Content has invalid size.  Was 367, should be 138.
Exception `JavaBuildpack::Util::Cache::InferredNetworkFailure' at /Users/danielmikusa/.rbenv/versions/2.2.3/lib/ruby/2.2.0/net/http.rb:1448 - Content has invalid size.  Was 367, should be 138.
```

You can see the request headers from the JBP permit gzip'd encoding and so the GH server is returning that.  The JBP does not seem to take this possibility into account when it's doing the content size check though.

Sample app to reproduce:

```
require 'java_buildpack/buildpack'
require 'java_buildpack/util/cache'
require 'java_buildpack/util/cache/download_cache'

JavaBuildpack::Buildpack.with_buildpack('/Users/danielmikusa/Downloads/java-buildpack/build_dir', 'Compile failed with exception %s') do |buildpack|
    JavaBuildpack::Util::Cache::DownloadCache.new().get('https://raw.githubusercontent.com/yooture/runtimes/master/index.yml') do |file|
          puts YAML.load_file(file)
    end
end
```

**Comment by nebhale** [comment]
Fixed in a3caeff96e0d6b8e5d4e22dd7557c68b73e55b89.

**Comment by thelangley** [comment]
can this please be merged into a new release? we ran into this issue today which we fixed by amending config as specified in commit a3caeff
